subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/FileCacheBackedScriptClassCompiler.java

```diff
@@ -61,74 +65,102 @@
-        final String dslId = operation.getId();
-        HashCode classLoaderHash = classLoaderHierarchyHasher.getClassLoaderHash(classLoader);
-        if (classLoaderHash == null) {
-            throw new IllegalArgumentException("Unknown classloader: " + classLoader);
-        }
-        final RemappingScriptSource remapped = new RemappingScriptSource(source);
-
-        PrimitiveHasher hasher = Hashing.newPrimitiveHasher();
-        hasher.putString(dslId);
-        hasher.putHash(sourceHashCode);
-        hasher.putHash(classLoaderHash);
-        String key = hasher.hash().toCompactString();
-
-        // Caching involves 2 distinct caches, so that 2 scripts with the same (hash, classpath) do not get compiled twice
-        // 1. First, we look for a cache script which (path, hash) matches. This cache is invalidated when the compile classpath of the script changes
-        // 2. Then we look into the 2d cache for a "generic script" with the same hash, that will be remapped to the script class name
-        // Both caches can be closed directly after use because:
-        // For 1, if the script changes or its compile classpath changes, a different directory will be used
-        // For 2, if the script changes, a different cache is used. If the classpath changes, the cache is invalidated, but classes are remapped to 1. anyway so never directly used
-        final PersistentCache cache = cacheBuilderFactory.createCacheBuilder("scripts/" + key)
-            .withDisplayName(dslId + " generic class cache for " + source.getDisplayName())
-            .withInitializer(new ProgressReportingInitializer(
-                progressLoggerFactory,
-                new CompileToCrossBuildCacheAction(remapped, classLoader, operation, verifier, scriptBaseClass),
-                "Compiling " + source.getShortDisplayName()))
-            .open();
-        try {
-            File genericClassesDir = classesDir(cache, operation);
-            File metadataDir = metadataDir(cache);
-            ClassPath remappedClasses = remapClasses(genericClassesDir, remapped);
-            return scriptCompilationHandler.loadFromDir(source, sourceHashCode, targetScope, remappedClasses, metadataDir, operation, scriptBaseClass);
-        } finally {
-            cache.close();
-        }
+        RemappingScriptSource remapped = new RemappingScriptSource(source);
+        ClassLoader classLoader = targetScope.getExportClassLoader();
+        File outputDir = doCompile(target, dslId, sourceHashCode, remapped, classLoader, operation, verifier, scriptBaseClass);
+
+        File genericClassesDir = classesDir(outputDir, operation);
+        File metadataDir = metadataDir(outputDir);
+        ClassPath remappedClasses = remapClasses(genericClassesDir, remapped);
+        return scriptCompilationHandler.loadFromDir(source, sourceHashCode, targetScope, remappedClasses, metadataDir, operation, scriptBaseClass);
```

Comment:
❓ This way we are still running the remapping outside of the execution engine. What's the plan with that? I can imagine a number of options:

1) Drop remapping altogether at a later data. We don't do it for Kotlin, so why do it for Groovy?
2) Include remapping in the unit of work. This would make the work non-relocatable, and thus something we shouldn't store in the build cache. I probably wouldn't do this.
3) Add another unit of work for the remapping, one that is not cacheable.

I think I would do either 1) or 3). Again, we don't have to do this now, but I think if we don't, then we should have a plan, and that plan should be a `TODO` here.